encryptionkeyrotation/README.md

```diff
@@ -0,0 +1,95 @@
+# CSI Addons Operation: EncryptionKeyRotation
+
+## Terminology
+
+| Term     | Definition                                                                            |
+| -------- | ------------------------------------------------------------------------------------- |
+| VolumeID | The identifier of the volume generated by the plugin.                                 |
+| CO       | Container Orchestration system that communicates with plugins using CSI service RPCs. |
+| SP       | Storage Provider, the vendor of a CSI plugin implementation.                          |
+| RPC      | [Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call).         |
+
+## Objective
+
+Define an extension to the CSI specification that will enable storage vendors
+(SP) to develop controllers/plugins that can rotate the Encryption Keys (KEK)
+of existing volumes that are encrypted.
+
+### Goals in MVP
+
+The new extension will define a procedure that
+
+- can be called for existing encrypted volumes
+- functions for Block mode volumes
+- interacts with the Node-Plugin
+- makes it possible for the SP to rotate encrpytion keys on volumes
+
+### Non-Goals in MVP
+
+- Rotating encrpytion keys for inactive volumes
+
+## Solution Overview
+
+This specification defines an interface along with the minimum operational and
+packaging recommendations for a storage provider (SP) to implement a key rotate operation for volumes. The interface declares the RPCs that a plugin
+MUST expose.
+
+## RPC Interface
+
+- **Node Service**: The Node plugin MAY implement this RPC.
+
+```protobuf
+syntax = "proto3";
+package encryptionkeyrotation;
+
+import "github.com/container-storage-interface/spec/lib/go/csi/csi.proto";
+import "google/protobuf/descriptor.proto";
+
+option go_package = "github.com/csi-addons/spec/lib/go/encryptionkeyrotation";
+
+// EncryptionKeyRotationController holds the RPC method for running
+// key rotation operation on the volume.
+service EncryptionKeyRotationController {
+    // EncryptionKeyRotate is a procedure that is called on the CSI Nodeplugin
+    rpc EncryptionKeyRotate(EncryptionKeyRotateRequest)
+    returns (EncryptionKeyRotateResponse){}
+}
+```
+
+### EncryptionKeyRotateRequest
```

Comment:
We only need the `volume_id`. Rest all is done on the node server end of things.

Reply:
Many CSI procedures include both the `volume_id` and a staging/target path. This procedure seems to be most usable on the Node-plugin once the volume is staged/published. To keep this procedure similar to others (like `NodeReclaimSpace`?), including a path would make sense, I think?

Reply:
Question, can we make this field optional? As in case of LUKS we need the device path. And getting the underline device path is driver specific. Making it an optional field would serve the cases where the path is required by the driver implementing the spec.

Reply:
yes, I think it can be optional. There might be storage providers that do not need to run the key rotation while the volume is staged/published.

I wonder if it makes sense to have `ControllerEncryptionKeyRotate` and `NodeEncryptionKeyRotate` for this. That way a CSI-driver can inform the caller what it supports, and the path is either not needed/available, or required.

Can you have a look at the ReclaimSpace procedures and see if that approach makes sense for EncryptionKeyRotation as well?

Reply:
Instead of going for `Controller` or `Node` specific RPCs I think having a common procedure is better suited as one would either process the key rotation on either the provisioner or nodeplugin, not both (most of the cases).

In the case of LUKS, we need the access to underlying device that is formatted for LUKS. Hence nodeplugin seems fitting.

In the cases where providers do not want to run the process while a volume is in use, they can register this service on the Provisioner instead. This approach leaves the implementation choice to the providers.

This is my understanding though, please feel free to add.

Reply:
Yes, that is an option too. If you want to go that route, make sure you explain both ways well in the README.md.

Reply:
Ack.